onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorQAttention.cpp

```diff
@@ -0,0 +1,656 @@
+        }
+        else
+        {
+            ML_CHECK_VALID_ARGUMENT(weightTensorShape[1] % 3 == 0);
+        }
+
+        const uint32_t hiddenSize = qkvHiddenSizes.empty() ? weightTensorShape[1] / 3 : qkvHiddenSizes[0];
+        const uint32_t vHiddenSize = qkvHiddenSizes.empty() ? weightTensorShape[1] / 3 : qkvHiddenSizes[2];
+        const uint32_t headSize = hiddenSize / numHeads;
+        const uint32_t vHeadSize = vHiddenSize / numHeads;
+        const uint32_t batchSize = inputTensorShape[0];
+        const uint32_t sequenceLength = inputTensorShape[1];
+
+        uint32_t desiredWeightTensorShape[3] = {batchSize, weightTensorShape[0], hiddenSize + hiddenSize + vHiddenSize};
+        MLOperatorTensorDataType dataType = kernelCreationContext.GetOutputEdgeDescription(outputIndex).tensorDataType;
+
+        m_inputTensorDescs[dmlWeightsIndex] = TensorDesc::ConstructBroadcastedTensorDesc(kernelCreationContext.GetInputEdgeDescription(weightsIndex).tensorDataType,
+                                                                                        desiredWeightTensorShape,
+                                                                                        weightTensorShape);
+        m_inputTensorDescs[inputScaleIndex] = TensorDesc::ConstructBroadcastedTensorDesc(kernelCreationContext.GetInputEdgeDescription(inputScaleIndex).tensorDataType,
+                                                                                        inputTensorShape,
+                                                                                        m_inputTensorDescs[inputScaleIndex].GetSizes());
+        m_inputTensorDescs[inputZeroPointIndex] = TensorDesc::ConstructBroadcastedTensorDesc(kernelCreationContext.GetInputEdgeDescription(inputZeroPointIndex).tensorDataType,
+                                                                                            inputTensorShape,
+                                                                                            m_inputTensorDescs[inputZeroPointIndex].GetSizes());
+        m_inputTensorDescs[weightScaleIndex] = TensorDesc::ConstructBroadcastedTensorDesc(kernelCreationContext.GetInputEdgeDescription(weightScaleIndex).tensorDataType,
+                                                                                            desiredWeightTensorShape,
+                                                                                            m_inputTensorDescs[weightScaleIndex].GetSizes());
+        m_inputTensorDescs[weightZeroPointIndex] = TensorDesc::ConstructBroadcastedTensorDesc(kernelCreationContext.GetInputEdgeDescription(weightZeroPointIndex).tensorDataType,
+                                                                                            desiredWeightTensorShape,
+                                                                                            m_inputTensorDescs[weightZeroPointIndex].GetSizes());
```

Comment:
nit: alignment is a bit weird and inconsistent here. The line is pretty long anyway, so it's probably better to start with the first parameter on the next line and only have a +4 indent instead of aligning it with the opening parentheses.

Reply:
Concur. Other parts of ORT vary arbitrarily between 2 spaces or 4 spaces or 93 spaces (e.g. if opening paren) with a complicated set of rules and exceptions to those rules that I've never quite figured out yet 😅, but standard indentation in the DML EP and DML code is consistently just simply 4 spaces (which is very easy to remember :)). Also this line is 178 columns wide, whereas we typically shoot for no more than ~120 (granted, there are some understandable exceptions, like the huge operator registration table where wrapping would worsen readability, but generally anyway).